rllib/examples/autoregressive_action_dist.py

```diff
@@ -0,0 +1,212 @@
+            obs_space, action_space, num_outputs, model_config, name)
+        if action_space != Tuple([Discrete(2), Discrete(2)]):
+            raise ValueError(
+                "This model only supports the [2, 2] action space")
+
+        # Inputs
+        obs_input = tf.keras.layers.Input(
+            shape=obs_space.shape, name="obs_input")
+        a1_input = tf.keras.layers.Input(shape=(1, ), name="a1_input")
+        ctx_input = tf.keras.layers.Input(
+            shape=(num_outputs, ), name="ctx_input")
+
+        # Output of the model (normally 'logits', but for an autoregressive
+        # dist this is more like a context/feature layer encoding the obs)
+        context = tf.keras.layers.Dense(
+            num_outputs,
+            name="hidden",
+            activation=tf.nn.tanh,
+            kernel_initializer=normc_initializer(1.0))(obs_input)
+
+        # V(s)
+        value_out = tf.keras.layers.Dense(
+            1,
+            name="value_out",
+            activation=None,
+            kernel_initializer=normc_initializer(0.01))(context)
+
+        # P(a1 | obs)
+        a1_logits = tf.keras.layers.Dense(
+            2,
+            name="a1_logits",
+            activation=None,
+            kernel_initializer=normc_initializer(0.01))(ctx_input)
+
+        # P(a2 | a1)
+        # --note: typically you'd want to implement P(a2 | a1, obs) as follows:
+        # a2_context = tf.keras.layers.Concatenate(axis=1)(
+        #     [ctx_input, a1_input])
+        a2_context = a1_input
+        a2_hidden = tf.keras.layers.Dense(
+            16,
+            name="a2_hidden",
+            activation=tf.nn.tanh,
+            kernel_initializer=normc_initializer(1.0))(a2_context)
+        a2_logits = tf.keras.layers.Dense(
+            2,
+            name="a2_logits",
+            activation=None,
+            kernel_initializer=normc_initializer(0.01))(a2_hidden)
+
+        # Base layers
+        self.base_model = tf.keras.Model(obs_input, [context, value_out])
+        self.register_variables(self.base_model.variables)
+        self.base_model.summary()
+
+        # Autoregressive action sampler
+        self.action_model = tf.keras.Model([ctx_input, a1_input],
+                                           [a1_logits, a2_logits])
+        self.action_model.summary()
+        self.register_variables(self.action_model.variables)
+
+    def forward(self, input_dict, state, seq_lens):
+        context, self._value_out = self.base_model(input_dict["obs"])
+        return context, state
+
+    def value_function(self):
+        return tf.reshape(self._value_out, [-1])
+
+
+if __name__ == "__main__":
+    ray.init()
+    args = parser.parse_args()
+    ModelCatalog.register_custom_model("autoregressive_model",
+                                       AutoregressiveActionsModel)
+    ModelCatalog.register_custom_action_dist("binary_autoreg_output",
+                                             BinaryAutoregressiveOutput)
```

Comment:
why make this explicit?

Reply:
We could auto register it but that would be more effort for sure.